### ISSUE TITLE
Incorrect error order for println!

### ISSUE THREAD

**Issue by pickfire** [rust-lang/rust#78568]
I tried this code:

```rust
fn main() {
    let guess = "123";
    let guess = match guess.trim().parse() {
        Ok(num) => num + 1,
        Err(_) => todo!(),
    };
    println!("guess: {}", guess);
}
```

I expected to see this happen:
```
  |
4 |         Ok(num) => num + 1,
  |            ^^^ cannot infer type
  |
help: consider specifying the type argument in the method call
  |
3 |     let guess = match guess.trim().parse::<F>() {
  |                                         ^^^^^
```

Instead, this happened:
```
  |
7 |     println!("guess: {}", guess);
  |                           ^^^^^ `()` cannot be formatted with the default formatter
  |
  = help: the trait `std::fmt::Display` is not implemented for `()`
  = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
  = note: required by `std::fmt::Display::fmt`
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

But eventually I will be able to see that once I use `{:?}` but users need to do an extra step to see the other warning.

### Meta

`rustc --version --verbose`: stable 1.47.0 to nightly 1.49.0-nightly (2020-10-29 6bdae9edd0cc099daa60)
```
stable 1.47.0 to nightly 1.49.0-nightly (2020-10-29 6bdae9edd0cc099daa60)
```

**Comment by estebank** [comment]
CC #38508.

**Comment by anoshyn** [comment]
Hi @niklasmarderx — are you still planning to work on this? I noticed the claim is from a couple months back with no linked PR, so I wanted to check in before stepping on anything in-flight. Happy to take it over if it's no longer on your plate.

**Comment by anoshyn** [comment]
Quick triage finding — on a stage1 build of current `main` (`rustc 1.97.0-dev`), this turns out to be **edition-gated by the never-type fallback** (RFC 3299), not a `format_args` issue.

Reproducing the example across editions:

| Edition | Diagnostic |
|---|---|
| 2015 / 2018 / 2021 | `error[E0277]: "()" doesn't implement Display` (the reported misleading behavior) |
| 2024 | `error[E0284]: type annotations needed` on `parse()` (the correct error) |

Smoking-gun: forcing the fallback the "wrong" way inverts the error.

```rust
#![allow(internal_features)]
#![feature(rustc_attrs, never_type)]
#![rustc_never_type_options(fallback = "never")]  // or "unit"
```

- 2021 + `fallback = "never"` → E0284 (helpful)
- 2024 + `fallback = "unit"` → E0277 (misleading)

**Mechanism.** The `Err(_) => todo!()` arm has type `!`. Pre-2024, `!` falls back to `()`, so the `match` unifies at `()` and silently absorbs the unresolved `T` from `parse()`, leaving `guess: ()` — the downstream Display error then becomes the only thing reported. Post-2024, `!` stays as `!`, the match keeps `T` ambiguous, and the real E0284 surfaces on `parse()` directly. The gate lives at `compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs:506-514` (`default_fallback`).

**Implication.** Any compiler-side fix amounts to changing the fallback in editions ≤2021, which is the breaking change RFC 3299 deliberately deferred to 2024. The issue is already fixed in edition 2024 and can't be cleanly fixed pre-2024 without a source-compat break.

Suggesting this could be closed as "fixed by edition-2024 never-type fallback (RFC 3299)" — but happy to leave open if there's appetite for a softer pre-2024 improvement (e.g. a note attached to E0277 when the `()` likely originated from `!`-to-`()` fallback).

**Comment by phtown** [comment]
This now works in all editions on nightly, scheduled for stabilization with 1.100. There's still a more general issue of never-type trait resolution being debated in #36375, but given !: Display we can probably close this issue.
